File: nmLabReport/lab20.c
#include <stdio.h>

void newtonForwardDifference(float x[], float y[], int n, float value) {
    float h = x[1] - x[0];
    float diffTable[n][n];
    float derivative;

    // Initialize the difference table
    for (int i = 0; i < n; i++) {
        diffTable[i][0] = y[i];
    }

    // Calculate the forward differences
    for (int j = 1; j < n; j++) {
        for (int i = 0; i < n - j; i++) {
            diffTable[i][j] = diffTable[i + 1][j - 1] - diffTable[i][j - 1];
        }
    }

    // Calculate the derivative using Newton's Forward Difference formula
    derivative = diffTable[0][1] / h;

    printf("The derivative at x = %.2f is %.2f\n", value, derivative);
}

int main() {
    int n;
    float value;

    printf("Enter the number of data points: ");
    scanf("%d", &n);

    float x[n], y[n];

    printf("Enter the data points (x and y):\n");
    for (int i = 0; i < n; i++) {
        printf("x[%d] = ", i);
        scanf("%f", &x[i]);
        printf("y[%d] = ", i);
        scanf("%f", &y[i]);
    }

    printf("Enter the value of x at which the derivative is to be calculated: ");
    scanf("%f", &value);

    newtonForwardDifference(x, y, n, value);

    return 0;
}

File: nmLabReport/lab20.py
def newton_forward_difference(x, y, n, value):
    h = x[1] - x[0]
    diff_table = [[0 for _ in range(n)] for _ in range(n)]

    # Initialize the difference table
    for i in range(n):
        diff_table[i][0] = y[i]

    # Calculate the forward differences
    for j in range(1, n):
        for i in range(n - j):
            diff_table[i][j] = diff_table[i + 1][j - 1] - diff_table[i][j - 1]

    # Calculate the derivative using Newton's Forward Difference formula
    derivative = diff_table[0][1] / h

    print(f"The derivative at x = {value:.2f} is {derivative:.2f}")

def main():
    n = int(input("Enter the number of data points: "))
    x = []
    y = []

    print("Enter the data points (x and y):")
    for i in range(n):
        x.append(float(input(f"x[{i}] = ")))
        y.append(float(input(f"y[{i}] = ")))

    value = float(input("Enter the value of x at which the derivative is to be calculated: "))

    newton_forward_difference(x, y, n, value)

if __name__ == "__main__":
    main()